commit 722683709c07503d9516181addd6c0597f8d9d84 (refs/changes/85/4585/4)
Author: Jared Morrow <jm@joyent.com>
Date:   2018-08-07T12:04:22-06:00 (1 year, 2 months ago)
    
    MANTA-3774 Use fastdelete method to delete objects if user account has snaplinks disabled
    Reviewed by: Jan Wyszynski <jan.wyszynski@joyent.com>
    Approved by: Jan Wyszynski <jan.wyszynski@joyent.com>

diff --git a/lib/moray.js b/lib/moray.js
index 7db1c9e..35f2981 100644
--- a/lib/moray.js
+++ b/lib/moray.js
@@ -35,7 +35,7 @@ var utils = require('./utils');
 var sprintf = util.format;
 
 var BUCKET = process.env.MANTA_RING_BUCKET || 'manta';
-var BUCKET_VERSION = 1;
+var BUCKET_VERSION = 2;
 /* JSSTYLED */
 var ROOT_RE = /^\/[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\/stor$/;
 var SCHEMA = {
@@ -305,18 +305,28 @@ function recordDeleteLog(req, cb) {
     }
     log.debug('object ' + prevObjectId + ' is candidate for deletion.');
 
-    //now log to the manta_delete_log table...
+    // now log to the manta_delete_log table or the manta_fastdelete_queue...
     var now = Math.round((microtime.now() / 1000));
     var _key = '/' + prevObjectId + '/' + now;
     var _value = JSON.stringify(prevmd);
     var _etag = crc.hex32(crc.crc32(_value));
     var _mtime = now;
-    var objectId = prevObjectId;
-    var sql = 'INSERT INTO manta_delete_log (_key, _value, _etag, ' +
-        '_mtime, objectId) VALUES ($1, $2, $3, $4, $5)';
-    var values = [_key, _value, _etag, _mtime, objectId];
+    var sql = '';
+    var values = [];
+
+    // If snaplinks are disabled use the fastdelete_queue rather than delete_log
+    if (req.headers['x-muskie-snaplinks-disabled']) {
+        log.debug('object ' + prevObjectId + ' being added to fastdelete.');
+        sql = 'INSERT INTO manta_fastdelete_queue (_key, _value, _etag, ' +
+            '_mtime) VALUES ($1, $2, $3, $4)';
+        values = [prevObjectId, _value, _etag, _mtime];
+    } else {
+        sql = 'INSERT INTO manta_delete_log (_key, _value, _etag, ' +
+            '_mtime, objectId) VALUES ($1, $2, $3, $4, $5)';
+        values = [_key, _value, _etag, _mtime, prevObjectId];
+    }
 
-    //execute
+    // execute
     var q = req.pg.query(sql, values);
     q.once('error', function (err) {
         log.debug(err, 'manta delete log insert: failed');
@@ -931,6 +941,8 @@ Moray.prototype.delMetadata = function delMetadata(options, callback) {
     assert.string(options.key, 'options.key');
     assert.string(options.requestId, 'options.requestId');
     assert.object(options, 'options.previousMetadata');
+    assert.optionalBool(options.snapLinksDisabled,
+        'options.snapLinksDisabled');
     assert.func(callback, 'callback');
 
     if (!this.client) {
@@ -946,12 +958,20 @@ Moray.prototype.delMetadata = function delMetadata(options, callback) {
     var log = this.log;
     var opts = {
         req_id: options.requestId,
-        etag: options._etag,
-        headers: {
-            'x-muskie-prev-metadata': options.previousMetadata
-        }
+        etag: options._etag
     };
 
+    if (options.snapLinksDisabled) {
+        opts.headers = {
+            'x-muskie-prev-metadata': options.previousMetadata,
+            'x-muskie-snaplinks-disabled': true
+        };
+    } else {
+        opts.headers = {
+            'x-muskie-prev-metadata':  options.previousMetadata
+        };
+    }
+
     log.debug({
         key: key,
         etag: opts.etag,
